File: src/orderbook.rs
use crate::types::{Order, OrderBook, Side};

impl OrderBook {
    pub fn add_limit_order(&mut self, order: Order) {
        if order.price >= self.bids.len() {
            return;
        }

        let book = match order.side {
            Side::Buy => &mut self.bids,
            Side::Sell => &mut self.asks,
        };

        let level = &mut book[order.price];

        level.total_qty += order.qty;
        level.orders.push_back(order.id);

        self.orders.insert(order.id, order);
    }

    pub fn execute_order(&mut self, order_id: u64, qty: u64) {
        if let Some(order) = self.orders.get_mut(&order_id) {
            let traded = qty.min(order.qty);

            order.qty -= traded;

            let book = match order.side {
                Side::Buy => &mut self.bids,
                Side::Sell => &mut self.asks,
            };

            let level = &mut book[order.price];

            level.total_qty -= traded;

            if order.qty == 0 {
                if let Some(pos) = level.orders.iter().position(|id| *id == order_id) {
                    level.orders.remove(pos);
                }

                self.orders.remove(&order_id);
            }
        }
    }

    pub fn cancel_order(&mut self, order_id: u64, qty: u64) {
        if let Some(order) = self.orders.get_mut(&order_id) {
            let cancel = qty.min(order.qty);

            order.qty -= cancel;

            let book = match order.side {
                Side::Buy => &mut self.bids,
                Side::Sell => &mut self.asks,
            };

            let level = &mut book[order.price];

            level.total_qty -= cancel;

            if order.qty == 0 {
                if let Some(pos) = level.orders.iter().position(|id| *id == order_id) {
                    level.orders.remove(pos);
                }

                self.orders.remove(&order_id);
            }
        }
    }

    pub fn delete_order(&mut self, order_id: u64) {
        if let Some(order) = self.orders.remove(&order_id) {
            let book = match order.side {
                Side::Buy => &mut self.bids,
                Side::Sell => &mut self.asks,
            };

            let level = &mut book[order.price];

            level.total_qty -= order.qty;

            if let Some(pos) = level.orders.iter().position(|id| *id == order_id) {
                level.orders.remove(pos);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::types::{Order, OrderBook, Side, MAX_PRICE};

    fn buy(id: u64, price: usize, qty: u64) -> Order {
        Order { id, price, qty, side: Side::Buy }
    }

    fn sell(id: u64, price: usize, qty: u64) -> Order {
        Order { id, price, qty, side: Side::Sell }
    }

    // ── add_limit_order ──────────────────────────────────────────────────────

    #[test]
    fn test_add_buy_order_updates_bids() {
        let mut book = OrderBook::new();
        book.add_limit_order(buy(1, 100, 50));
        assert_eq!(book.bids[100].total_qty, 50);
        assert_eq!(book.bids[100].orders.len(), 1);
        assert_eq!(book.bids[100].orders[0], 1);
        assert!(book.orders.contains_key(&1));
    }

    #[test]
    fn test_add_sell_order_updates_asks() {
        let mut book = OrderBook::new();
        book.add_limit_order(sell(2, 200, 30));
        assert_eq!(book.asks[200].total_qty, 30);
        assert_eq!(book.asks[200].orders.len(), 1);
        assert_eq!(book.asks[200].orders[0], 2);
        assert!(book.orders.contains_key(&2));
    }

    #[test]
    fn test_add_buy_does_not_touch_asks() {
        let mut book = OrderBook::new();
        book.add_limit_order(buy(1, 100, 50));
        assert_eq!(book.asks[100].total_qty, 0);
        assert!(book.asks[100].orders.is_empty());
    }

    #[test]
    fn test_add_order_price_at_max_is_ignored() {
        let mut book = OrderBook::new();
        let order = Order { id: 99, price: MAX_PRICE, qty: 10, side: Side::Buy };
        book.add_limit_order(order);
        assert!(!book.orders.contains_key(&99));
    }

    #[test]
    fn test_add_multiple_orders_same_price_accumulates_qty() {
        let mut book = OrderBook::new();
        book.add_limit_order(buy(1, 100, 50));
        book.add_limit_order(buy(2, 100, 30));
        assert_eq!(book.bids[100].total_qty, 80);
        assert_eq!(book.bids[100].orders.len(), 2);
    }

    #[test]
    fn test_add_orders_fifo_queue_order() {
        let mut book = OrderBook::new();
        book.add_limit_order(buy(1, 100, 10));
        book.add_limit_order(buy(2, 100, 20));
        book.add_limit_order(buy(3, 100, 30));
        let orders = &book.bids[100].orders;
        assert_eq!(orders[0], 1);
        assert_eq!(orders[1], 2);
        assert_eq!(orders[2], 3);
    }

    // ── execute_order ────────────────────────────────────────────────────────

    #[test]
    fn test_execute_partial_reduces_qty() {
        let mut book = OrderBook::new();
        book.add_limit_order(buy(1, 100, 50));
        book.execute_order(1, 20);
        assert_eq!(book.orders[&1].qty, 30);
        assert_eq!(book.bids[100].total_qty, 30);
        assert_eq!(book.bids[100].orders.len(), 1);
    }

    #[test]
    fn test_execute_full_removes_order() {
        let mut book = OrderBook::new();
        book.add_limit_order(buy(1, 100, 50));
        book.execute_order(1, 50);
        assert!(!book.orders.contains_key(&1));
        assert_eq!(book.bids[100].total_qty, 0);
        assert!(book.bids[100].orders.is_empty());
    }

    #[test]
    fn test_execute_more_than_qty_clamps_and_removes() {
        let mut book = OrderBook::new();
        book.add_limit_order(buy(1, 100, 50));
        book.execute_order(1, 100);
        assert!(!book.orders.contains_key(&1));
        assert_eq!(book.bids[100].total_qty, 0);
    }

    #[test]
    fn test_execute_sell_order() {
        let mut book = OrderBook::new();
        book.add_limit_order(sell(1, 150, 80));
        book.execute_order(1, 40);
        assert_eq!(book.orders[&1].qty, 40);
        assert_eq!(book.asks[150].total_qty, 40);
    }

    #[test]
    fn test_execute_nonexistent_order_is_noop() {
        let mut book = OrderBook::new();
        book.execute_order(999, 10); // must not panic
    }

    #[test]
    fn test_execute_removes_from_price_level_queue() {
        let mut book = OrderBook::new();
        book.add_limit_order(buy(1, 100, 10));
        book.add_limit_order(buy(2, 100, 20));
        book.execute_order(1, 10);
        let orders = &book.bids[100].orders;
        assert_eq!(orders.len(), 1);
        assert_eq!(orders[0], 2);
        assert_eq!(book.bids[100].total_qty, 20);
    }

    // ── cancel_order ─────────────────────────────────────────────────────────

    #[test]
    fn test_cancel_partial_reduces_qty() {
        let mut book = OrderBook::new();
        book.add_limit_order(sell(1, 150, 80));
        book.cancel_order(1, 30);
        assert_eq!(book.orders[&1].qty, 50);
        assert_eq!(book.asks[150].total_qty, 50);
        assert_eq!(book.asks[150].orders.len(), 1);
    }

    #[test]
    fn test_cancel_full_removes_order() {
        let mut book = OrderBook::new();
        book.add_limit_order(sell(1, 150, 80));
        book.cancel_order(1, 80);
        assert!(!book.orders.contains_key(&1));
        assert_eq!(book.asks[150].total_qty, 0);
        assert!(book.asks[150].orders.is_empty());
    }

    #[test]
    fn test_cancel_more_than_qty_clamps_and_removes() {
        let mut book = OrderBook::new();
        book.add_limit_order(sell(1, 150, 80));
        book.cancel_order(1, 200);
        assert!(!book.orders.contains_key(&1));
        assert_eq!(book.asks[150].total_qty, 0);
    }

    #[test]
    fn test_cancel_buy_order() {
        let mut book = OrderBook::new();
        book.add_limit_order(buy(1, 100, 60));
        book.cancel_order(1, 20);
        assert_eq!(book.orders[&1].qty, 40);
        assert_eq!(book.bids[100].total_qty, 40);
    }

    #[test]
    fn test_cancel_nonexistent_order_is_noop() {
        let mut book = OrderBook::new();
        book.cancel_order(999, 10); // must not panic
    }

    // ── delete_order ─────────────────────────────────────────────────────────

    #[test]
    fn test_delete_buy_order() {
        let mut book = OrderBook::new();
        book.add_limit_order(buy(1, 100, 50));
        book.delete_order(1);
        assert!(!book.orders.contains_key(&1));
        assert_eq!(book.bids[100].total_qty, 0);
        assert!(book.bids[100].orders.is_empty());
    }

    #[test]
    fn test_delete_sell_order() {
        let mut book = OrderBook::new();
        book.add_limit_order(sell(5, 200, 100));
        book.delete_order(5);
        assert!(!book.orders.contains_key(&5));
        assert_eq!(book.asks[200].total_qty, 0);
        assert!(book.asks[200].orders.is_empty());
    }

    #[test]
    fn test_delete_removes_from_queue_leaves_others() {
        let mut book = OrderBook::new();
        book.add_limit_order(buy(1, 100, 10));
        book.add_limit_order(buy(2, 100, 20));
        book.delete_order(1);
        assert_eq!(book.bids[100].orders.len(), 1);
        assert_eq!(book.bids[100].orders[0], 2);
        assert_eq!(book.bids[100].total_qty, 20);
    }

    #[test]
    fn test_delete_nonexistent_order_is_noop() {
        let mut book = OrderBook::new();
        book.delete_order(999); // must not panic
    }

    // ── mixed operations ─────────────────────────────────────────────────────

    #[test]
    fn test_mixed_add_execute_cancel_delete() {
        let mut book = OrderBook::new();
        book.add_limit_order(buy(1, 100, 100));
        book.add_limit_order(buy(2, 100, 50));
        book.execute_order(1, 60);
        book.cancel_order(2, 25);
        book.delete_order(1);
        assert!(!book.orders.contains_key(&1));
        assert_eq!(book.orders[&2].qty, 25);
        assert_eq!(book.bids[100].total_qty, 25);
        assert_eq!(book.bids[100].orders.len(), 1);
    }
}

File: cpp/orderbook.cpp
#include "orderbook.h"
#include <algorithm>

// ============================================================
// Constructor
// ============================================================

OrderBook::OrderBook()
    : bids_(MAX_PRICE),  // heap-allocated; default PriceLevel{} per slot
      asks_(MAX_PRICE)
{}

// ============================================================
// Order management
// ============================================================

void OrderBook::add_limit_order(const Order& order) {
    if (order.price == 0 || order.price >= MAX_PRICE) return; // guard

    auto& book  = (order.side == Side::Buy) ? bids_ : asks_;
    auto& level = book[order.price];

    level.total_qty += order.qty;
    level.orders.push_back(order.id);   // FIFO: new orders go to the back
    orders_[order.id] = order;
}

void OrderBook::execute_order(uint64_t order_id, uint32_t qty) {
    auto it = orders_.find(order_id);
    if (it == orders_.end()) return;

    Order& order   = it->second;
    uint32_t traded = (qty < order.qty) ? qty : order.qty; // min
    order.qty -= traded;

    auto& book  = (order.side == Side::Buy) ? bids_ : asks_;
    auto& level = book[order.price];
    level.total_qty -= static_cast<uint64_t>(traded);

    if (order.qty == 0) {
        // Remove the ID from the FIFO queue.
        auto& q   = level.orders;
        auto  pos = std::find(q.begin(), q.end(), order_id);
        if (pos != q.end()) q.erase(pos);
        orders_.erase(it);
    }
}

void OrderBook::cancel_order(uint64_t order_id, uint32_t qty) {
    // A cancel is a partial or full quantity reduction — identical to execute.
    execute_order(order_id, qty);
}

void OrderBook::delete_order(uint64_t order_id) {
    auto it = orders_.find(order_id);
    if (it == orders_.end()) return;

    const Order& order = it->second;
    auto& book  = (order.side == Side::Buy) ? bids_ : asks_;
    auto& level = book[order.price];

    // Saturating subtract to avoid underflow on corrupted state.
    if (level.total_qty >= static_cast<uint64_t>(order.qty))
        level.total_qty -= order.qty;
    else
        level.total_qty = 0;

    auto& q   = level.orders;
    auto  pos = std::find(q.begin(), q.end(), order_id);
    if (pos != q.end()) q.erase(pos);

    orders_.erase(it);
}

// ============================================================
// Drain methods (used once per batch by the matcher)
// ============================================================

std::vector<Order> OrderBook::drain_buys() {
    std::vector<Order> result;
    // Iterate price levels from high to low (natural BUY priority order).
    for (int p = static_cast<int>(MAX_PRICE) - 1; p >= 0; --p) {
        auto& level = bids_[p];
        while (!level.orders.empty()) {
            uint64_t id = level.orders.front();
            level.orders.pop_front();
            auto it = orders_.find(id);
            if (it != orders_.end()) {
                result.push_back(it->second);
                orders_.erase(it);
            }
        }
        level.total_qty = 0;
    }
    return result;
}

std::vector<Order> OrderBook::drain_asks() {
    std::vector<Order> result;
    // Iterate price levels from low to high (natural SELL priority order).
    for (uint32_t p = 0; p < MAX_PRICE; ++p) {
        auto& level = asks_[p];
        while (!level.orders.empty()) {
            uint64_t id = level.orders.front();
            level.orders.pop_front();
            auto it = orders_.find(id);
            if (it != orders_.end()) {
                result.push_back(it->second);
                orders_.erase(it);
            }
        }
        level.total_qty = 0;
    }
    return result;
}

// ============================================================
// Reporting helpers (non-destructive)
// ============================================================

std::vector<const Order*> OrderBook::resting_buys_sorted() const {
    std::vector<const Order*> result;
    result.reserve(orders_.size());
    for (const auto& [id, order] : orders_) {
        if (order.side == Side::Buy) result.push_back(&order);
    }
    std::sort(result.begin(), result.end(),
        [](const Order* a, const Order* b) {
            if (a->price     != b->price)     return a->price     > b->price;     // DESC
            if (a->timestamp != b->timestamp) return a->timestamp < b->timestamp; // ASC
            return a->id < b->id;                                                  // ASC
        });
    return result;
}

std::vector<const Order*> OrderBook::resting_asks_sorted() const {
    std::vector<const Order*> result;
    result.reserve(orders_.size());
    for (const auto& [id, order] : orders_) {
        if (order.side == Side::Sell) result.push_back(&order);
    }
    std::sort(result.begin(), result.end(),
        [](const Order* a, const Order* b) {
            if (a->price     != b->price)     return a->price     < b->price;     // ASC
            if (a->timestamp != b->timestamp) return a->timestamp < b->timestamp; // ASC
            return a->id < b->id;                                                  // ASC
        });
    return result;
}
